wagtail/contrib/search_promotions/views.py

```diff
@@ -1,85 +1,64 @@
-        ordering = request.GET["ordering"]
-
-    # Query
-    queries = Query.objects.filter(editors_picks__isnull=False).distinct()
-
-    if "views" in ordering:
-        queries = queries.annotate(views=functions.Coalesce(Sum("daily_hits__hits"), 0))
-
-    queries = queries.order_by(ordering)
-
-    # Search
-    is_searching = False
-    query_string = request.GET.get("q", "")
-
-    if query_string:
-        queries = queries.filter(query_string__icontains=query_string)
-        is_searching = True
-
-    # Paginate
-    paginator = Paginator(queries, per_page=20)
-    try:
-        queries = paginator.page(request.GET.get("p", 1))
-    except InvalidPage:
-        raise Http404
-
-    if request.headers.get("x-requested-with") == "XMLHttpRequest":
-        return TemplateResponse(
-            request,
-            "wagtailsearchpromotions/results.html",
-            {
-                "is_searching": is_searching,
-                "ordering": ordering,
-                "queries": queries,
-                "query_string": query_string,
-            },
-        )
-    else:
-        return TemplateResponse(
-            request,
-            "wagtailsearchpromotions/index.html",
-            {
-                "is_searching": is_searching,
-                "ordering": ordering,
-                "queries": queries,
-                "query_string": query_string,
-                "search_form": SearchForm(
-                    data={"q": query_string} if query_string else None,
-                ),
-            },
+class IndexView(generic.IndexView):
+    model = Query
+    template_name = "wagtailsearchpromotions/index.html"
+    results_template_name = "wagtailsearchpromotions/results.html"
+    page_title = _("Search Terms")
+    any_permission_required = ["add", "change", "delete"]
+    paginate_by = 20
+    page_kwarg = "p"
+    index_url_name = "wagtailsearchpromotions:index"
+    is_searchable = True
+    default_ordering = "query_string"
+    columns = [
+        TitleColumn(
+            "query_string",
+            classname="query_string",
+            label=_("Search term(s)"),
+            width="40%",
+            url_name="wagtailsearchpromotions:edit",
+            sort_key="query_string",
+        ),
+        Column(
+            "promoted_results",
+            label=_("Promoted results"),
+            width="40%",
+        ),
+        Column("views", label=_("Views (Past week)"), width="20%", sort_key="views"),
+    ]
+
+    def get_base_queryset(self):
+        queryset = self.model.objects.filter(editors_picks__isnull=False).distinct()
+        queryset = queryset.annotate(
+            views=functions.Coalesce(Sum("daily_hits__hits"), 0),
+            promoted_results=F("editors_picks__page__title"),
         )
+        return queryset
+
+    def get_context_data(self, *args, **kwargs):
+        context = super().get_context_data(*args, **kwargs)
+        context["queries"] = self.get_queryset()
```

Comment:
We might be able to define `context_object_name = "queries"` on the view class and remove this altogether? Otherwise we'll be calling `get_queryset()` twice.